com.unity.netcode.gameobjects/Runtime/Messaging/Messages/RpcMessage.cs

```diff
@@ -19,54 +19,54 @@ public struct HeaderData
         }
 
         public HeaderData Header;
-        public FastBufferWriter RpcData;
+        public FastBufferWriter RpcWriteData;
+        internal FastBufferReader RpcReadData;
+        internal NetworkObject ReceivingNetworkObject;
+        internal NetworkBehaviour ReceivingNetworkBehaviour;
 
 
         public unsafe void Serialize(FastBufferWriter writer)
         {
-            if (!writer.TryBeginWrite(FastBufferWriter.GetWriteSize(Header) + RpcData.Length))
+            if (!writer.TryBeginWrite(FastBufferWriter.GetWriteSize(Header) + RpcWriteData.Length))
             {
                 throw new OverflowException("Not enough space in the buffer to store RPC data.");
             }
             writer.WriteValue(Header);
-            writer.WriteBytes(RpcData.GetUnsafePtr(), RpcData.Length);
+            writer.WriteBytes(RpcWriteData.GetUnsafePtr(), RpcWriteData.Length);
         }
 
-        public static void Receive(FastBufferReader reader, in NetworkContext context)
+        public bool Deserialize(FastBufferReader reader, ref NetworkContext context)
         {
-            var message = new RpcMessage();
-            if (!reader.TryBeginRead(FastBufferWriter.GetWriteSize(message.Header)))
+            if (!reader.TryBeginRead(FastBufferWriter.GetWriteSize(Header)))
             {
                 throw new OverflowException("Not enough space in the buffer to read RPC data.");
             }
-            reader.ReadValue(out message.Header);
-            message.Handle(reader, context, (NetworkManager)context.SystemOwner, context.SenderId, true);
+            reader.ReadValue(out Header);
+
+            var networkManager = (NetworkManager)context.SystemOwner;
+            if (!networkManager.SpawnManager.SpawnedObjects.ContainsKey(Header.NetworkObjectId))
+            {
+                networkManager.SpawnManager.TriggerOnSpawn(Header.NetworkObjectId, reader, ref context);
```

Comment:
huh?

Reply:
Not opposed to this change but I do think it should be a separate PR.